h5py/_hl/selections.py

```diff
@@ -91,6 +91,68 @@ def select(shape, args, dsid):
     return sel
 
 
+class MultiBlockSlice(object):
+
+    """
+        A conceptual extension of the built-in slice object to allow selections
+        using start, stride, count and block.
+
+        If given, these parameters will be passed directly to
+        H5Sselect_hyperslab. The defaults are start=0, stride=1, block=1,
+        count=length, which will select the full extent.
+
+        __init__(start, count, stride, block) => Create a new MultiBlockSlice, storing
+            any given selection parameters and using defaults for the others
+        start => The offset of the starting element of the specified hyperslab
+        count => The number of blocks to select
+        stride => The number of elements between the start of one block and the next
+        block => The number of elements in each block
+
+    """
+
+    def __init__(self, start=0, count=None, stride=1, block=1):
+        if start < 0:
+            raise ValueError("Start must be positive")
+        if count is not None and count < 1 or stride < 1 or block < 1:
+            raise ValueError("Count, block and stride must be non-zero")
+        if block > stride:
+            raise ValueError("Blocks will overlap if block > stride")
+
+        self.start = start
+        self.count = count
+        self.stride = stride
+        self.block = block
+
+    def indices(self, length):
+        """Calculate and validate start, stop, stride and block for the given length"""
+        if self.count is None:
+            # Select as many full blocks as possible without exceeding extent
+            count = (length - self.start - self.block) // self.stride + 1
+            if count < 1:
+                raise ValueError(
+                    "No full blocks can be selected using {} "
+                    "on dimension of length {}".format(self._repr(), length)
+                )
+        else:
+            count = self.count
+
+        end_index = self.start + self.block + (count - 1) * self.stride - 1
+        if end_index >= length:
+            raise ValueError(
+                "{} range ({} - {}) extends beyond maximum index ({})".format(
+                    self._repr(count), self.start, end_index, length - 1
+                ))
+
+        return self.start, count, self.stride, self.block
+
+    def _repr(self, count=None):
+        if count is None:
+            count = self.count
+        return "{}(start={}, stride={}, count={}, block={})".format(
+            self.__class__.__name__, self.start, self.stride, count, self.block
+        )
```

Comment:
Why not use this as an actual `__repr__` method? It may be handy to inspect these objects.

Reply:
I didn't use `__repr__` because I didn't know if it was good practice to define a __repr__ that takes an argument. I wanted it to take an argument in indices because there I have the calculated count to display, rather than `None`, and I thought that would be useful. What are your thoughts?

Reply:
Well, I added a parameter to `Dataset.__getitem__`, so I'm not in a position to complain about that. :wink: But probably a nicer way to do it is to leave this method as is, and define `__repr__` as a wrapper around it.